pkg/apis/autoscaling/v1/types.go

```diff
@@ -0,0 +1,118 @@
+	// e.g. 70 means that 70% of the requested CPU should be in use.
+	TargetPercentage int32 `json:"targetPercentage"`
+}
+
+// specification of a horizontal pod autoscaler.
+type HorizontalPodAutoscalerSpec struct {
+	// reference to Scale subresource; horizontal pod autoscaler will learn the current resource consumption from its status,
+	// and will set the desired number of pods by modifying its spec.
+	ScaleRef SubresourceReference `json:"scaleRef"`
+	// lower limit for the number of pods that can be set by the autoscaler, default 1.
+	MinReplicas *int32 `json:"minReplicas,omitempty"`
+	// upper limit for the number of pods that can be set by the autoscaler; cannot be smaller than MinReplicas.
+	MaxReplicas int32 `json:"maxReplicas"`
+	// target average CPU utilization (represented as a percentage of requested CPU) over all the pods;
+	// if not specified it defaults to the target CPU utilization at 80% of the requested resources.
+	CPUUtilization *CPUTargetUtilization `json:"cpuUtilization,omitempty"`
+}
+
+// current status of a horizontal pod autoscaler
+type HorizontalPodAutoscalerStatus struct {
+	// most recent generation observed by this autoscaler.
+	ObservedGeneration *int64 `json:"observedGeneration,omitempty"`
+
+	// last time the HorizontalPodAutoscaler scaled the number of pods;
+	// used by the autoscaler to control how often the number of pods is changed.
+	LastScaleTime *unversioned.Time `json:"lastScaleTime,omitempty"`
+
+	// current number of replicas of pods managed by this autoscaler.
+	CurrentReplicas int32 `json:"currentReplicas"`
+
+	// desired number of replicas of pods managed by this autoscaler.
+	DesiredReplicas int32 `json:"desiredReplicas"`
+
+	// current average CPU utilization over all pods, represented as a percentage of requested CPU,
+	// e.g. 70 means that an average pod is using now 70% of its requested CPU.
+	CurrentCPUUtilizationPercentage *int32 `json:"currentCPUUtilizationPercentage,omitempty"`
+}
+
+// configuration of a horizontal pod autoscaler.
+type HorizontalPodAutoscaler struct {
+	unversioned.TypeMeta `json:",inline"`
+	// Standard object metadata. More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	v1.ObjectMeta `json:"metadata,omitempty"`
+
+	// behaviour of autoscaler. More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status.
+	Spec HorizontalPodAutoscalerSpec `json:"spec,omitempty"`
+
+	// current information about the autoscaler.
+	Status HorizontalPodAutoscalerStatus `json:"status,omitempty"`
+}
+
+// list of horizontal pod autoscaler objects.
+type HorizontalPodAutoscalerList struct {
+	unversioned.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	unversioned.ListMeta `json:"metadata,omitempty"`
+
+	// list of horizontal pod autoscaler objects.
+	Items []HorizontalPodAutoscaler `json:"items"`
+}
+
+// ListOptions is the query options to a standard REST list call.
+// Copy-pasted from api/v1.
+type ListOptions struct {
```

Comment:
No, we are not using unversioned for these objects anymore.  We are using
versioned objects, and we are not moving things into unversioned/ for
params for things like this yet.

On Thu, Feb 4, 2016 at 8:28 AM, Jordan Liggitt <notifications@github.com>
wrote:

> In pkg/apis/autoscaling/v1/types.go
> <https://github.com/kubernetes/kubernetes/pull/20501#discussion_r51871188>
> :
>
> > +	Status HorizontalPodAutoscalerStatus `json:"status,omitempty"`
> > +}
> > +
> > +// list of horizontal pod autoscaler objects.
> > +type HorizontalPodAutoscalerList struct {
> > +	unversioned.TypeMeta `json:",inline"`
> > +	// Standard list metadata.
> > +	unversioned.ListMeta `json:"metadata,omitempty"`
> > +
> > +	// list of horizontal pod autoscaler objects.
> > +	Items []HorizontalPodAutoscaler `json:"items"`
> > +}
> > +
> > +// ListOptions is the query options to a standard REST list call.
> > +// Copy-pasted from api/v1.
> > +type ListOptions struct {
>
> v1 or unversioned? I thought we were trying to change dependencies on v1
> utility objects like this. @lavalamp <https://github.com/lavalamp>,
> @deads2k <https://github.com/deads2k>, does using v1 here seem right to
> you?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20501/files#r51871188>.
>